### PR TITLE
feat(sffv): Add `client.searchForFacetValues()` 

### PR DESCRIPTION
**Pull request by francoischalifour**
_This PR adds the top-level `client.searchForFacetValues([{ indexName, params }, ...queries])` method._

## Why

As we're going towards a more backend agnostic InstantSearch, we'd like to polish the internal JS client API to make it easier to write and understand.

With the current API, we need to initialize an index and to request SFFV on this index. This is an implementation detail that doesn't need to be exposed to our users.

Since AlgoliaSearch Transformers leverage this JS client interface to allow creation of custom backend implementations, we need to make it smoother. `searchForFacetValues()` should be at the same level as `search()`. For consistency, we also need to support multiple facet values queries (take an array of queries as parameter instead of a single query).

Right now, if you want to use a custom implementation backend that supports SFFV, you need to know the implementation:

```javascript
client.initIndex(indexName)
  .searchForFacetValues(params)
  .then(({ facetHits }) => {
      res.status(200).send(facetHits);
  });
```

## Proposal

Let's skip this `initIndex(indexName)` method and pass the `indexName` in the query object of the new `client.searchForFacetValues([{ indexName, params }, ...queries])` method.

The new usage will be easier and more powerful/flexible:

```javascript
client.searchForFacetValues([{ indexName, params }])
  .then(contents => {
    const [firstContent] = contents;
    const { facetHits } = firstContent;

    res.status(200).send(facetHits);
  });
```

## Backward compatibility

We should make our helper use this new method by default but fall back to the old one if the client used doesn't expose it. I will make a PR for this!

----

~_In this series of improvements, we'll make the helper use the promise version of `search(queries)` soon._~ ✅

### CODE REVIEW

**Comment by francoischalifour** [comment]
The Angular bundle size is [120B bigger than anticipated](https://bundlesize-store.now.sh/build?info=%7B%22files%22%3A%5B%7B%22maxSize%22%3A18432%2C%22path%22%3A%22.%2Fdist%2Falgoliasearch.angular.min.js%22%2C%22size%22%3A18265%7D%2C%7B%22maxSize%22%3A18432%2C%22path%22%3A%22.%2Fdist%2Falgoliasearch.jquery.min.js%22%2C%22size%22%3A18128%7D%2C%7B%22maxSize%22%3A47104%2C%22path%22%3A%22.%2Fdist%2Falgoliasearch.angular.js%22%2C%22size%22%3A47229%7D%2C%7B%22maxSize%22%3A47104%2C%22path%22%3A%22.%2Fdist%2Falgoliasearch.jquery.js%22%2C%22size%22%3A46810%7D%2C%7B%22maxSize%22%3A47104%2C%22path%22%3A%22.%2Fdist%2Falgoliasearch.parse.js%22%2C%22size%22%3A27745%7D%2C%7B%22maxSize%22%3A13312%2C%22path%22%3A%22.%2Fdist%2FalgoliasearchLite.min.js%22%2C%22size%22%3A12532%7D%2C%7B%22maxSize%22%3A32768%2C%22path%22%3A%22.%2Fdist%2FalgoliasearchLite.js%22%2C%22size%22%3A32377%7D%5D%2C%22repo%22%3A%22algolia%2Falgoliasearch-client-javascript%22%2C%22branch%22%3A%22feat%2Fsffv-top-level%22%2C%22commit_message%22%3A%22Merge%20994c742d67c1d43f73bebf599d08db239553b043%20into%20f7730e52bdd52e4e93706de3c57aa4a395f53a62%22%2C%22sha%22%3A%22994c742d67c1d43f73bebf599d08db239553b043%22%7D), which is why the build fails. We can increase the bundle size threshold from 46kB to 46.5kB if you're OK with that.

**Comment by vvo** [comment]
Thanks!

Just to be clear, this is right now needed to make it easier for us internally to create fake clients from transformers, is that right?

Good way to also raise the fact that we have a dual API in all API clients where everything under index.* could maybe just be sitting in client.* with indexName being available for some of those methods as a parameter.

cc @julienbourdeau worth investigating while reviewing the APIs of APIs clients

**Comment by francoischalifour** [comment]
> Just to be clear, this is right now needed to make it easier for us internally to create fake clients from transformers, is that right?

It is actually more than that. Let's say you want to implement Elasticsearch on a custom backend with InstantSearch like [this example](https://github.com/algolia/algoliasearch-transformers/blob/707e6982470d6ce3be3e639b718e93e31ad82797/packages/elasticsearch-transformer/examples/concerts-server/index.js#L58). You'd need to use `search()` and `initIndex.searchForFacetValues()` (instead of `searchForFacetValues()`). With this new API method, you could skip the "pointless" `initIndex()` call (from the user point of view).

To sum up, it would simplify the surface of the API that the user needs to know to use a custom backend implementation via a transformer.

Does that make sense?

**Comment by Haroenv** [comment]
If you're not using Algolia, why would you use the Algolia client backend?

**Comment by francoischalifour** [comment]
@Haroenv Are you referring to the `concert-server` example? This file has different endpoints to showcase all the possibilities. I'm not using `algoliasearch` for the Elasticsearch endpoints, but for the custom Algolia endpoint.

**Comment by bobylito** [comment]
Obviously I'm in favor of that PR, since I discussed it with @francoischalifour beforehand.

After discussing it with @Haroenv who had some reserve about it, there is a slight problem. This feature pushes for simplicity and symmetry. However this does not have complete symmetry with search. It is not able to do a multi index SFFV.

And that sounds like an actual good idea since we have a use case for this: a smart search box that would be able to complete facet values over multiple facetted attributes. Implementing the multi index search for facet values would be an enabler of such new UI pattern.

**Comment by vvo** [comment]
Agreed on implementing .searchForFacetValues([{indexName, ...}]).

About the example on https://github.com/algolia/algoliasearch-transformers/blob/707e6982470d6ce3be3e639b718e93e31ad82797/packages/elasticsearch-transformer/examples/concerts-server/index.js#L58, I understand now why it's easier to have searchForFacetValues at the root for such usecase where you really want to have the transformer acting on the backend and not on the frontend.

**Comment by francoischalifour** [comment]
I'm going to update this PR to pass an array of queries to `searchForFacetValues()`. It turns out it's not an easy task because of the testing environment current limitations.

**Comment by francoischalifour** [comment]
@Haroenv This is too painful to unit test this new feature because of the current way we deal with unit tests (our custom framework). I've tested the feature with a integration test and it works fine (@bobylito approved verbally).

Let me know what you think!

File: test/run-integration.js [hunk 1]
```diff
@@ -164,6 +165,27 @@ function searchForFacetValues(t) {
     .then(noop, _.bind(t.error, t));
 }
 
+function searchForFacetValues(t) {
+  t.plan(2);
+
+  client
+    .searchForFacetValues([
+      {
```
Comment (Haroenv) [inline test/run-integration.js]: would maybe be interesting to test one with a single query only too, because I imagine that will be used more

File: src/AlgoliaSearchCore.js [hunk 2]
```diff
@@ -609,6 +609,70 @@ AlgoliaSearchCore.prototype.search = function(queries, opts, callback) {
...
+  }
+
+  var client = this;
+
+  return Promise.all(map(queries, function performQuery(query) {
+    if (
+      !query ||
+      query.indexName === undefined ||
+      query.params.facetName === undefined ||
+      query.params.facetQuery === undefined
```
Comment (samouss) [inline src/AlgoliaSearchCore.js]: I'm not following the all picture of this discussion but why do we nest the `facetName`, `facetQuery` and the rest of the parameters in a `params` object? The signature of the function will be a bit less verbose.

```js
client.searchForFacetValues([
  {
    indexName: "products",
    params: {
      facetName: "brand",
      facetValue: "Apple"
      ...searchParameters,
    }
  }
]);

// vs

client.searchForFacetValues([
  {
    indexName: "products",
    facetName: "brand",
    facetValue: "Apple",
    ...searchParameters,
  }
]);
```

Comment (Haroenv) [reply src/AlgoliaSearchCore.js]: agreed with the second syntax, we use this kind of pattern in other multi-index methods too

Comment (bobylito) [reply src/AlgoliaSearchCore.js]: That's a good point if it matches the other method signatures.

Comment (samouss) [reply src/AlgoliaSearchCore.js]: I didn't look correctly to the `search()` signature of the client. It looks the same as the one implemented in the PR.

https://github.com/algolia/algoliasearch-client-javascript/blob/2533c50a9588f5eccff1ebcf9716f476451d9f60/src/AlgoliaSearchCore.js#L564-L578

Comment (francoischalifour) [reply src/AlgoliaSearchCore.js]: It's the same signature as `search()` indeed. Although I agree `params` is not necessary.

File: test/run-integration.js (hunk 1 above)
Comment (francoischalifour) [reply test/run-integration.js]: Thanks, I will add the test!
